drivers/timer/apic_timer.c

```diff
@@ -31,193 +27,63 @@ BUILD_ASSERT(!IS_ENABLED(CONFIG_SMP), "APIC timer doesn't support SMP");
- *
- * One important invariant that must be observed: `total_cycles` + `cached_icr`
- * is always an integral multiple of CYCLE_PER_TICK; this is, timer interrupts
- * are only ever scheduled to occur at tick boundaries.
- */
-
-static struct k_spinlock lock;
-static uint64_t total_cycles;
-static uint32_t cached_icr = CYCLES_PER_TICK;
-
-#ifdef CONFIG_TICKLESS_KERNEL
-
-static uint64_t last_announcement;	/* last time we called sys_clock_announce() */
-
-void sys_clock_set_timeout(int32_t n, bool idle)
-{
-	ARG_UNUSED(idle);
-
-	uint32_t ccr;
-	int   full_ticks;	/* number of complete ticks we'll wait */
-	uint32_t full_cycles;	/* full_ticks represented as cycles */
-	uint32_t partial_cycles;	/* number of cycles to first tick boundary */
-
-	if (n < 1) {
-		full_ticks = 0;
-	} else if ((n == K_TICKS_FOREVER) || (n > MAX_TICKS)) {
-		full_ticks = MAX_TICKS - 1;
-	} else {
-		full_ticks = n - 1;
-	}
-
-	full_cycles = full_ticks * CYCLES_PER_TICK;
-
-	/*
-	 * There's a wee race condition here. The timer may expire while
-	 * we're busy reprogramming it; an interrupt will be queued at the
-	 * local APIC and the ISR will be called too early, roughly right
-	 * after we unlock, and not because the count we just programmed has
-	 * counted down. Luckily this situation is easy to detect, which is
-	 * why the ISR actually checks to be sure the CCR is 0 before acting.
-	 */
-
-	k_spinlock_key_t key = k_spin_lock(&lock);
-
-	ccr = x86_read_loapic(LOAPIC_TIMER_CCR);
-	total_cycles += (cached_icr - ccr);
-	partial_cycles = CYCLES_PER_TICK - (total_cycles % CYCLES_PER_TICK);
-	cached_icr = full_cycles + partial_cycles;
-	x86_write_loapic(LOAPIC_TIMER_ICR, cached_icr);
-
-	k_spin_unlock(&lock, key);
-}
-
-uint32_t sys_clock_elapsed(void)
-{
-	uint32_t ccr;
-	uint32_t ticks;
-
-	k_spinlock_key_t key = k_spin_lock(&lock);
-	ccr = x86_read_loapic(LOAPIC_TIMER_CCR);
-	ticks = total_cycles - last_announcement;
-	ticks += cached_icr - ccr;
-	k_spin_unlock(&lock, key);
-	ticks /= CYCLES_PER_TICK;
-
-	return ticks;
-}
-
-static void isr(const void *arg)
-{
-	ARG_UNUSED(arg);
-
-	uint32_t cycles;
-	int32_t ticks;
-
-	k_spinlock_key_t key = k_spin_lock(&lock);
-
-	/*
-	 * If we get here and the CCR isn't zero, then this interrupt is
-	 * stale: it was queued while sys_clock_set_timeout() was setting
-	 * a new counter. Just ignore it. See above for more info.
-	 */
-
-	if (x86_read_loapic(LOAPIC_TIMER_CCR) != 0) {
-		k_spin_unlock(&lock, key);
-		return;
-	}
-
-	/* Restart the timer as early as possible to minimize drift... */
-	x86_write_loapic(LOAPIC_TIMER_ICR, MAX_TICKS * CYCLES_PER_TICK);
-
-	cycles = cached_icr;
-	cached_icr = MAX_TICKS * CYCLES_PER_TICK;
-	total_cycles += cycles;
-	ticks = (total_cycles - last_announcement) / CYCLES_PER_TICK;
-	last_announcement = total_cycles;
-	k_spin_unlock(&lock, key);
-	sys_clock_announce(ticks);
-}
+BUILD_ASSERT(CYCLES_PER_TICK >= 1, "APIC timer: bad CYCLES_PER_TICK");
```

Comment:
Might want a BUILD_ASSERT(!defined(CONFIG_TICKLESS_KERNEL)) given that support is being removed, for the benefit of any out-of-tree apps that might be using it.